File: python/31-40/Valid Sudoku.py
class Solution:
    # @param board, a 9x9 2D array
    # @return a boolean
    def isValidSudoku(self, board):
        
        #check row
        for row in board:
            record=[0 for i in range(0,9)]
            for cell in row:
                if cell =='.':
                    continue
                if record[int(cell)-1]!=1:
                    record[int(cell)-1] =1
                else:
                    return False
            record=[]

        #check col
        for j in range(0,9):
            record=[0 for k in range(0,9)]
            for i in range(0,9):
                cell = board[i][j]
                if cell =='.':
                    continue
                if record[int(cell)-1]!=1:
                    record[int(cell)-1]=1
                else:
                    return False
            record =[]
        #check sqar
        for i in range(0,9,3):
            for j in range(0,9,3):
                record=[0 for k in range(0,9)]
                for row in range(i,i+3):
                    for col in range(j,j+3):
                        cell = board[row][col]
                        if cell =='.':
                            continue
                        if record[int(cell)-1]!=1:
                            record[int(cell)-1]=1
                        else:
                            return False
                record=[]
        return True

File: Cpp/31-40/Valid Sudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        //check row , col ,square;
        unordered_set<int> rowSet,colSet;
        for(int i=0; i<9; i++){
          for(int j=0; j<9; j++){
            if(board[i][j]=='.')  continue;
            if(colSet.find(board[i][j])!=colSet.end())
              return false;
            colSet.insert(board[i][j]);
          }
          colSet.clear();
        }
        //check row
        for(int i=0; i<9; i++){
          for(int j=0; j<9; j++){
            if(board[j][i]=='.')  continue;
            if(rowSet.find(board[j][i])!=rowSet.end())
              return false;
            rowSet.insert(board[j][i]);
          }
          rowSet.clear();
        }
        //check square
        for(int i=0; i<3; i++)
          for(int j=0; j<3; j++){
            for(int row = i*3; row<i*3+3; row++)
              for(int col = j*3; col <j*3+3; col++){
                if(board[row][col]=='.')  continue;
                if(colSet.find(board[row][col])!=colSet.end())
                  return false;
                colSet.insert(board[row][col]);
              }
          colSet.clear();  
          }
          return true;
    }
};
